File: Graph/building-teams.py
from collections import defaultdict

def main():
    n, m = map(int, input().split())
    arr = [[int(x) for x in input().split()] for _ in range(m)]

    adj = defaultdict(list)
    for a1, a2 in arr:
        adj[a1].append(a2)
        adj[a2].append(a1)

    color = [-1] * (n+1)

    def isBipartite(node, color):
        stack = [(node, 1)]

        while stack:
            current, current_color = stack.pop()

            if color[current] == -1:
                color[current] = current_color

            for neighbor in adj[current]:
                if color[neighbor] == -1:
                    color[neighbor] = 3 - current_color
                    stack.append((neighbor, 3 - current_color))
                elif color[neighbor] == color[current]:
                    return False

        return True

    for i in range(1, n+1):
        if not isBipartite(i, color):
            print("IMPOSSIBLE")
            return

    print(' '.join([str(x) for x in color[1:]]))



if __name__=='__main__':
    main()

File: Graph/building-teams.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

bool isBipartite(int node, vector<int>& color, unordered_map<int, vector<int>>& adj, bool& result) {
    if (color[node] == -1) {
        color[node] = 1;
    }

    for (int it : adj[node]) {
        if (color[it] == -1) {
            color[it] = 3 - color[node];
            if(!isBipartite(it, color, adj, result)){
                cout << "IMPOSSIBLE" << endl;
                exit(0);
            }
        } else if (color[it] == color[node]) {
            result = false;
            return false;
        }
    }
    return true;
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<vector<int>> arr(m, vector<int>(2));
    unordered_map<int, vector<int>> adj;

    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < 2; ++j) {
            cin >> arr[i][j];
            adj[arr[i][j]].push_back(arr[i][(j + 1) % 2]);
        }
    }

    vector<int> color(n + 1, -1);

    bool result = true;
    for (int i = 1; i <= n; ++i) {
        if (color[i] == -1) {
            if(!isBipartite(i, color, adj, result)){
                cout << "IMPOSSIBLE" << endl;
                exit(0);
            }
        }
    }

    
    for (int i = 1; i <= n; ++i) {
        cout << color[i] << ' ';
        
        cout << endl;
    }

    return 0;
}
